File: onyx-database/src/main/java/com/onyx/descriptor/RelationshipDescriptor.java
package com.onyx.descriptor;

import com.onyx.persistence.annotations.CascadePolicy;
import com.onyx.persistence.annotations.FetchPolicy;
import com.onyx.persistence.annotations.RelationshipType;

/**
 * Created by timothy.osborn on 12/11/14.
 *
 * Details on an entity relationship
 */
public class RelationshipDescriptor extends AbstractBaseDescriptor
{
    public RelationshipDescriptor(){
        super();
    }

    private RelationshipType relationshipType;
    @SuppressWarnings("WeakerAccess")
    protected String inverse;
    @SuppressWarnings("WeakerAccess")
    protected Class inverseClass;
    private Class parentClass;
    @SuppressWarnings("WeakerAccess")
    protected FetchPolicy fetchPolicy;
    @SuppressWarnings("WeakerAccess")
    protected CascadePolicy cascadePolicy;
    @SuppressWarnings("WeakerAccess")
    protected EntityDescriptor entityDescriptor;

    @SuppressWarnings("WeakerAccess")
    protected byte loadFactor = 1;

    /**
     * This method is to determine what scale the underlying structure should be.  The values are from 1-10.
     * 1 is the fastest for small data sets.  10 is to span huge data sets intended that the performance of the index
     * does not degrade over time.  Note: You can not change this ad-hoc.  You must re-build the index if you intend
     * to change.  Always plan for scale when designing your data model.
     *
     * @param loadFactor Value from 1-10.
     * @since 1.2.0
     */
    public void setLoadFactor(byte loadFactor) {
        this.loadFactor = loadFactor;
    }

    /**
     * Getter for the load factor
     *
     * @return The values are from 1-10.
     * <p>
     * 1 is the fastest for small data sets.  10 is to span huge data sets intended that the performance of the index
     * does not degrade over time.  Note: You can not change this ad-hoc.  You must re-build the index if you intend
     * to change.  Always plan for scale when designing your data model.
     * @since 1.2.0
     */
    public byte getLoadFactor() {
        return this.loadFactor;
    }

    public RelationshipType getRelationshipType()
    {
        return relationshipType;
    }

    void setRelationshipType(RelationshipType relationshipType)
    {
        this.relationshipType = relationshipType;
    }

    public String getInverse()
    {
        return inverse;
    }

    public void setInverse(String inverse)
    {
        this.inverse = inverse;
    }

    public Class getInverseClass()
    {
        return inverseClass;
    }

    public void setInverseClass(Class inverseClass)
    {
        this.inverseClass = inverseClass;
    }

    public Class getParentClass()
    {
        return parentClass;
    }

    void setParentClass(Class parentClass)
    {
        this.parentClass = parentClass;
    }

    public FetchPolicy getFetchPolicy()
    {
        return fetchPolicy;
    }

    public void setFetchPolicy(FetchPolicy fetchPolicy)
    {
        this.fetchPolicy = fetchPolicy;
    }

    public CascadePolicy getCascadePolicy()
    {
        return cascadePolicy;
    }

    public void setCascadePolicy(CascadePolicy cascadePolicy)
    {
        this.cascadePolicy = cascadePolicy;
    }

    public EntityDescriptor getEntityDescriptor()
    {
        return entityDescriptor;
    }

    public void setEntityDescriptor(EntityDescriptor entityDescriptor)
    {
        this.entityDescriptor = entityDescriptor;
    }

    ////////////////////////////////////////////////////////
    //
    //  Hashing Object Overrides
    //
    ////////////////////////////////////////////////////////
    /**
     * Used for calculating hash structure
     *
     * @return Hash Code of a relationship
     */
    @Override
    public int hashCode()
    {
        if(entityDescriptor.getPartition() != null)
        {
            return (getParentClass().getName() + getName() + getInverseClass().getName() + getInverse() + entityDescriptor.getPartition().getPartitionValue()).hashCode();
        }
        return (getParentClass().getName() + name + getInverseClass().getName() + getInverse()).hashCode();
    }

    /**
     * Used for usage within a hashmap
     * @param val Value to compare
     * @return Whether the parameter is equal to this
     */
    @Override
    public boolean equals(Object val)
    {
        if(val instanceof RelationshipDescriptor)
        {
            RelationshipDescriptor comparison = (RelationshipDescriptor) val;
            if(entityDescriptor.getPartition() != null && comparison.getEntityDescriptor().getPartition() != null)
            {
                return (((RelationshipDescriptor) val).getParentClass().getName().equals(getParentClass().getName())
                        && ((RelationshipDescriptor) val).getName().equals(getName())
                        && ((RelationshipDescriptor) val).getInverseClass().getName().equals(getInverseClass().getName())
                        && ((RelationshipDescriptor) val).getInverse().equals(getInverse())
                        && entityDescriptor.getPartition().getPartitionValue().equals(comparison.getEntityDescriptor().getPartition().getPartitionValue())
                        && comparison.loadFactor == this.loadFactor);
            }
            else if(entityDescriptor.getPartition() != null)
            {
                return false;
            }
            else if(comparison.getEntityDescriptor().getPartition() != null)
            {
                return false;
            }

            return (((RelationshipDescriptor) val).getParentClass().getName().equals(getParentClass().getName())
            && ((RelationshipDescriptor) val).getName().equals(getName())
            && ((RelationshipDescriptor) val).getInverseClass().getName().equals(getInverseClass().getName())
            && ((RelationshipDescriptor) val).getInverse().equals(getInverse())
            && comparison.loadFactor == this.loadFactor);
        }

        return false;
    }
}

File: onyx-database/src/main/kotlin/com/onyx/descriptor/RelationshipDescriptor.kt
package com.onyx.descriptor

import com.onyx.extension.common.ClassMetadata
import com.onyx.persistence.annotations.values.CascadePolicy
import com.onyx.persistence.annotations.values.FetchPolicy
import com.onyx.persistence.annotations.values.RelationshipType
import kotlin.jvm.internal.Intrinsics

/**
 * Created by timothy.osborn on 12/11/14.
 *
 * Details on an entity relationship
 */
data class RelationshipDescriptor(
        var inverse: String? = null,
        var parentClass: Class<*>,
        var relationshipType: RelationshipType,
        var inverseClass: Class<*>,
        var fetchPolicy: FetchPolicy,
        var cascadePolicy: CascadePolicy,
        var name: String = "",
        var type: Class<*> = ClassMetadata.ANY_CLASS
) : AbstractBaseDescriptor() {

    lateinit var entityDescriptor: EntityDescriptor

    override fun hashCode(): Int = ((((((((if (this.inverse != null) this.inverse!!.hashCode() else 0) * 31 + this.parentClass.hashCode()) * 31 + this.relationshipType.hashCode()) * 31 + this.inverseClass.hashCode()) * 31 + this.fetchPolicy.hashCode()) * 31 + this.cascadePolicy.hashCode()) * 31) * 31 + this.name.hashCode()) * 31 + this.type.hashCode()

    override fun equals(other: Any?): Boolean {
        if (this !== other) {
            if (other !is RelationshipDescriptor) {
                return false
            }

            val var2 = other as RelationshipDescriptor?
            if (!Intrinsics.areEqual(this.inverse, var2!!.inverse) || !Intrinsics.areEqual(this.parentClass, var2.parentClass) || !Intrinsics.areEqual(this.relationshipType, var2.relationshipType) || !Intrinsics.areEqual(this.inverseClass, var2.inverseClass) || !Intrinsics.areEqual(this.fetchPolicy, var2.fetchPolicy) || !Intrinsics.areEqual(this.cascadePolicy, var2.cascadePolicy) || !Intrinsics.areEqual(this.name, var2.name) || !Intrinsics.areEqual(this.type, var2.type)
                    && Intrinsics.areEqual(this.entityDescriptor.partition, var2.entityDescriptor.partition)) {
                return false
            }
        }

        return true
    }

    val isToMany:Boolean
        get() = relationshipType === RelationshipType.MANY_TO_MANY || relationshipType === RelationshipType.ONE_TO_MANY

    val isToOne:Boolean
        get() = !isToMany

    val shouldSaveEntity:Boolean
        get() = cascadePolicy === CascadePolicy.SAVE || cascadePolicy === CascadePolicy.ALL

    val shouldDeleteEntity:Boolean
        get() = (cascadePolicy === CascadePolicy.DELETE || cascadePolicy === CascadePolicy.ALL) && isToOne

    val shouldDeleteEntityReference:Boolean
        get() = ((cascadePolicy === CascadePolicy.DELETE || cascadePolicy === CascadePolicy.ALL) && isToMany) || isToOne

}
